src/k8s/pkg/k8sd/app/cluster_util.go

```diff
@@ -61,6 +63,20 @@ func waitApiServerReady(ctx context.Context, snap snap.Snap) error {
 	return nil
 }
 
+func waitControlPlaneServices(ctx context.Context, snap snap.Snap) error {
+	// The services may be able to start, appearing to be "active", but they might eventually fail due to
+	// various reasons, and they may be restarted. We're checking their activity a few times.
+	if err := control.Consistently(ctx, 3, 5*time.Second, func() error {
+		if err := snaputil.CheckControlPlaneServicesStates(ctx, snap, "active"); err != nil {
+			return fmt.Errorf("failed to ensure all control plane services are active: %w", err)
+		}
```

Comment:
I'm not the biggest fan of adding 15 seconds to every bootstrap...

Reply:
I am not a fan of any kind of hard-coded time. What about this?

> Is it not possible to poll the state frequently and continue as soon as all services are active?

Reply:
Not really, systemd reports a unit as `active` immediately after starting it, even if it'll fail almost immediately afterwards. Ofcourse, it'll eventually mark it as failed afterwards, and it'll try again to start it due to the restart policy.

If you are to check the unit state immediately after starting it, you'll most likely see it as `active`. We can easily test this. Let's say we have a ``--foo=lish`` argument in ``/var/snap/k8s/common/args/kubelet``, which will cause ``kubelet`` to fail. If we then run the following command:

```
svc_name="snap.k8s.kubelet.service" && sudo systemctl stop $svc_name && sudo systemctl start $svc_name && systemctl is-active $svc_name && sleep 5 && systemctl is-active $svc_name
```

We'd see the following:

```
active
failed
```

If the sleep is shorter (2s), you might even see it in `activating` state.

An idea would be to have `ExecStartPost` in the systemd units, which could help ensure that the service is running. As an interesting tidbit, it seems that setting `ExecStartPost=/usr/bin/sleep 1` would actually protect against immediately-exiting services (though I'm not 100% it is an intended behaviour).